tests/tests/client.rs

```diff
@@ -0,0 +1,208 @@
+use self::support::*;
+extern crate tokio_connect;
+
+use h2_support::{mock::Mock, prelude::*};
+use tokio::executor::current_thread::*;
+
+use tower_h2::client::Connect;
+
+use tower_service::{NewService, Service};
+use futures::future::{self, FutureResult};
+use std::cell::RefCell;
+
+mod support;
+
+struct MockConn {
+    conn: RefCell<Option<Mock>>,
+}
+
+impl MockConn {
+    fn new(mock: Mock) -> Self {
+        MockConn {
+            conn: RefCell::new(Some(mock))
+        }
+    }
+}
+
+impl tokio_connect::Connect for MockConn {
+    type Connected = Mock;
+    type Error = ::std::io::Error;
+    type Future = FutureResult<Mock, ::std::io::Error>;
+
+    fn connect(&self) -> Self::Future {
+        future::ok(self.conn.borrow_mut().take().expect("connected more than once!"))
+    }
+}
+
+#[test]
+fn hello() {
+    let _ = ::env_logger::try_init();
+
+    let (io, srv) = mock::new();
+
+    let srv = srv
+        .assert_client_handshake()
+        .unwrap()
+        .recv_settings()
+        .recv_frame(
+            frames::headers(1)
+                .request("GET", "https://example.com/")
+                .eos(),
+        )
+        .send_frame(frames::headers(1).response(200).eos())
+        .close();
+
+    let conn = MockConn::new(io);
+    let h2 = Connect::new(conn, Default::default(), TaskExecutor::current());
+
+    let done = h2.new_service()
+        .map_err(|e| panic!("connect err: {:?}", e))
+        .and_then(|mut h2| {
+            h2.call(http::Request::builder()
+                .method("GET")
+                .uri("https://example.com/")
+                .body(())
+                .unwrap())
+        })
+        .map(|rsp| {
+            assert_eq!(rsp.status(), http::StatusCode::OK);
+        })
+        .map_err(|e| panic!("error: {:?}", e));
+
+    CurrentThread::new()
+        .spawn(srv)
```

Comment:
I don't think that's the case. I did a quick test by adding a `recv_frame` call for a frame that isn't sent, so that the mock expects it instead of the frame that is sent, the test fails (as it should):

```diff
diff --git a/tests/tests/client.rs b/tests/tests/client.rs
index 516c7ed..78d3d7b 100644
--- a/tests/tests/client.rs
+++ b/tests/tests/client.rs
@@ -44,6 +44,7 @@ fn hello() {
         .assert_client_handshake()
         .unwrap()
         .recv_settings()
+        .recv_frame(frames::data(9000, "barf"))
         .recv_frame(
             frames::headers(1)
                 .request("GET", "https://example.com/")
```
results in:
```
    Finished dev [unoptimized + debuginfo] target(s) in 4.57s
     Running target/debug/deps/client-16c5e429aba4bcf5

running 4 tests
test hello ... FAILED
test hello_rsp_body ... ok
test hello_bodies ... ok
test hello_req_body ... ok

failures:

---- hello stdout ----
	thread 'hello' panicked at 'assertion failed: `(left == right)`
  left: `Some(Frame::Headers(Headers { stream_id: StreamId(1), stream_dep: None, flags: HeadersFlag { end_stream: true, end_headers: true, padded: false, priority: false } }))`,
 right: `Some(Frame::Data(Data { stream_id: StreamId(9000), flags: DataFlags, pad_len: None }))`: recv_frame', /Users/eliza/.cargo/git/checkouts/h2-5627e19d2c95dbee/2b960b8/tests/h2-support/src/mock.rs:674:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.


failures:
    hello

test result: FAILED. 3 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--test client'
```

Reply:
Ok, this is an issue w/ the current-thread executor then.

Reply:
I filed https://github.com/tokio-rs/tokio/issues/640 to track.